tokio-io/src/async_write.rs

```diff
@@ -172,17 +172,20 @@ impl AsyncWrite for std_io::Sink {
     }
 }
 
-// TODO: Implement `prepare_uninitialized_buffer` for `io::Take`.
-// This is blocked on rust-lang/rust#27269
 impl<T: AsyncRead> AsyncRead for std_io::Take<T> {
+    unsafe fn prepare_uninitialized_buffer(&self, buf: &mut [u8]) -> bool {
+        self.get_ref().prepare_uninitialized_buffer(buf)
+    }
 }
 
-// TODO: Implement `prepare_uninitialized_buffer` when upstream exposes inner
-// parts
 impl<T, U> AsyncRead for std_io::Chain<T, U>
     where T: AsyncRead,
           U: AsyncRead,
 {
+    unsafe fn prepare_uninitialized_buffer(&self, buf: &mut [u8]) -> bool {
+        let (t, u) = self.get_ref();
+        t.prepare_uninitialized_buffer(buf) & u.prepare_uninitialized_buffer(buf)
```

Comment:
You only need to call the second `prepare_uninitialized_buffer` if the first returns `false:

> If this function returns true, then the memory has been zeroed out. This allows implementations of AsyncRead which are composed of multiple sub implementations to efficiently implement prepare_uninitialized_buffer.

Reply:
That said, this is a good example of what is wrong with this API.

Reply:
Indeed.  I don't know what I've been thinking. :)